### ISSUE TITLE
Log something only once (or only N times)

### ISSUE THREAD

**Issue by MatthewScholefield** [Delgan/loguru#383]
I have a situation where I want to log something only once. It would be nice to have some `logger.opt(once=True)`. For example:

```python
def fetch_document(...):  # Repeatedly called
    document = database.retrieve(...)
    if document.endswith(r'\r\n'):
        logger.opt(once=True).warning(r'Database is storing \r\n  endings. Please reconfigure.')
```

Without such an option, either a non-fatal warning bloats the logs repeating the same message, or I'd need to unpleasantly use a global variable:

```python
did_log_database_warning = False
def fetch_document(...):
    global did_log_database_warning
    ...
    if document.endswith(r'\r\n') and not did_log_database_warning:
        did_log_database_warning = True
        logger.warning(...)
```

Naturally, this would lead to me writing a help function, `log_once`. So, I figured it would make sense to include something like this with loguru. What do you think?

To support the `log_once = logger.opt(once=True); log_once('here'); log_once('there')`, it could store a set of printed log statements based on the line number of the logger call. Alternatively, using a new method, it could simply store it once per instance like `logger.limit(1).warning('hello')` which would implicitly allow another use case of constraining the number of lines of some related set of warnings ie.:

```python
io_logger = logger.limit(100, overflow_msg='Suppressing future io errors')
def decrypt_chars(data):
    for offset, c in enumerate(data):
        if cannot_decrypt(c):
            io_logger.warning('Could not decrypt byte {} at offset {}', hex(c), offset)
            continue
        d = decrypt(c)
        if is_invalid_char(d):
            io_logger.warning('Invalid character {!r} at offset {}', d, offset)
        yield d
```

**Comment by Delgan** [comment]
Hi @MatthewScholefield. Sorry for replying so late.

This seems to be a very interesting idea. I remember at some point I was also looking for a way to prevent excessive logging of the same message. Finally I did not implement anything because I was not sure such rate-limiting for logging was good practice or not.

I wish I could just extend `opt()` but it seems too restrictive. We probably need to suppress logs for a period of time only (eg. no more than 100 calls every minute), which means more attributes are needed. The proposed `.limit()` method looks good.

Alternatively, one could use another library like [`ratelimiter`](https://pypi.org/project/ratelimiter). It should work perfectly well. The downside is that it require wrapping logging call with a context manager, which is surely less convenient that using  `logger.limit()` in-line.

So, yeah, thanks for your suggestion! It could be worth integrating it into Loguru, I will keep thinking about it.

**Comment by liz-zither** [comment]
I'm quite interested in this feature as well. Can you expand on the `ratelimiter` idea with some pseudo code?  I'm not sure how to implement.

**Comment by Delgan** [comment]
@liz-zither You can use the `ratelimiter` library like that:

```python
from ratelimiter import RateLimiter

for i in range(100):
    with RateLimiter(max_calls=10, period=1):
        logger.info("Iteration #{}", i)
```

**Comment by Delgan** [comment]
@liz-zither Ok, my bad. The `ratelimiter` library is definitely not designed for this use case. Sorry, I didn't look into the details.

Here is another example using a custom class:

```python
from loguru import logger
from unittest.mock import MagicMock


class LimitedLogger:
    def __init__(self, limit=None):
        self._limit = limit
        self._count = 0

    def get(self):
        if self._limit is None or self._count < self._limit:
            self._count += 1
            return logger
        else:
            return MagicMock()


limited_logger = LimitedLogger(limit=10)

for i in range(100):
    limited_logger.get().info("Iteration #{}", i)
```

**Comment by DeflateAwning** [comment]
It seems that the documentation on this feature got missed. Can someone write the final usage of this feature here? Perhaps we can also add it to the [Help Guide](https://loguru.readthedocs.io/en/stable/resources.html).

**Comment by Delgan** [comment]
@DeflateAwning It's not yet documented because it's not yet implemented. :)

Actually I don't know if this should be provided as built-in in Loguru or as a third-library.

**Comment by cdgiv** [comment]
I'd love this feature as well! I've worked with internal logging systems before that offered a `log_every_N()` type method which, not quite the same as what's requested here, but I think is flexible enough to satisfy several needs.

**Comment by ice-tong** [comment]
Same feature request here!

**Comment by Moreh-LeeJunhyeok** [comment]
also requests this feature. log_N_times is useful in simple warning cases

**Comment by drdarina** [comment]
Same feature request!

**Comment by noamgot** [comment]
I came up with this simple POC:

```python
import atexit as _atexit
import sys as _sys
from collections import defaultdict

from loguru import _defaults
from loguru._logger import Core as _Core
from loguru._logger import Logger as _Logger


class MyLogger(_Logger):
    _msg_history = defaultdict(set)

    def _log(self, level, from_decorator, options, message, args, kwargs):
        once = kwargs.pop("once", False)
        if once:
            if message in self._msg_history[level]:
                return
            self._msg_history[level].add(message)
        super()._log(level, from_decorator, options, message, args, kwargs)


logger = MyLogger(
    core=_Core(),
    exception=None,
    depth=0,
    record=False,
    lazy=False,
    colors=False,
    raw=False,
    capture=True,
    patchers=[],
    extra={},
)
if _defaults.LOGURU_AUTOINIT and _sys.stderr:
    logger.add(_sys.stderr)

_atexit.register(logger.remove)

logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message 123", once=True)


logger.info("This is a warning message", once=True)
logger.info("This is a warning message", once=True)

```
seems to work just fine:
![image](https://github.com/user-attachments/assets/0e99fa39-842b-4af6-959e-918930af9510)

Asking the core devs here - do you think it can be plugged into the current implementation?

**Comment by Delgan** [comment]
@noamgot It looks like you can use the existing `filter` parameter to implement such solution, without requiring to touch Loguru internals:

```python
import sys
from collections import defaultdict

from loguru import logger

_msg_history = defaultdict(set)


def filter(record):
    if "once" in record["extra"]:
        level = record["level"].no
        message = record["message"]
        if message in _msg_history[level]:
            return False
        _msg_history[level].add(message)
    return True


logger.remove()
logger.add(sys.stderr, filter=filter)

logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message 123", once=True)


logger.info("This is a warning message", once=True)
logger.info("This is a warning message", once=True)
```

**Comment by noamgot** [comment]
Looks like a decent solution, works for me. thanks @Delgan

**Comment by SzymonWojciechowskiLimango** [comment]
Hey, I was trying to make it done, but I am getting an error. My case is slightly more complicated, but still - trivial I guess.
```{python}
a = {"a", "b", "c"}
b = {"a"}

logger.info(f"New set: {a - b}", once=True)
```
I get the following error:
```KeyError: "'c', 'b'"```
Looks like the `{a-b}` part has not been evaluated prior to passing to the logger. I am using the 0.7.3 version.

**Comment by wjaskowski** [comment]
@SzymonWojciechowskiLimango Since this problem hit me just yesterday, I can explain: if there is any extra parameter (like `once=True`, then loguru turns on its internal formatting system, which starts to interpret `{...}` expressions.

Thus, a workaround is to use:
```
logger.info("New set: {}", a - b, once=True)
```
or
```
logger.info("New set: {exp}", exp=a - b, once=True)
```

**Comment by DeflateAwning** [comment]
Template strings in Python 3.14 will solve this perfectly (perhaps pending a bit more implementation here)!

**Comment by Delgan** [comment]
@SzymonWojciechowskiLimango @wjaskowski This an unfortunate problem that may arise while combining f-strings with keyword arguments. See some explanations and workaround here: [Why logging a message with f-string sometimes raises an exception?](https://loguru.readthedocs.io/en/stable/resources/troubleshooting.html#why-logging-a-message-with-f-string-sometimes-raises-an-exception).

I plan to change the API in a future version of Loguru to avoid this kind of unexpected errors. Template strings might be part of the solution.

**Comment by liblaf** [comment]
@Delgan's solution is decent. However, there may be a minor issue. If a filter is used by two handlers, records will only be logged by one of them. This occurs because when the second handler processes the data, it calls `filter()` a second time with the same `record`, which has already been recorded in the `history`. One solution is to allocate a separate `history` to each handler. I've also considered an alternative approach: caching the filter's output. Since different handlers feed identical `record`s into the filter, we can determine record identity based on `id(record)`. I'm not sure if this is elegant, but at least it avoids allocating multiple history sets.

<details>

<summary>Code (click to expand)</summary>

```python
import sys

import loguru
from loguru import logger


class FilterOnce:
    history: set
    _cache: tuple["loguru.Record | None", bool]

    def __init__(self) -> None:
        self.history = set()
        self._cache = (None, False)

    def __call__(self, record: "loguru.Record") -> bool:
        once = record["extra"].get("once", False)
        if not once:
            return True
        if record is self._cache[0]:
            return self._cache[1]
        key = (record["level"].no, record["message"])
        if key in self.history:
            self._cache = (record, False)
            return False
        self.history.add(key)
        self._cache = (record, True)
        return True


logger.remove()
filter_once = FilterOnce()
logger.add(sys.stderr, filter=filter_once)
logger.add("once.log", filter=filter_once)

logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message 123", once=True)

logger.info("This is a warning message", once=True)
logger.info("This is a warning message", once=True)
```

</summary>

**Comment by Delgan** [comment]
Thank you for your insight, @liblaf. I agree that the snippet I provided was shortsighted and actually incorrect.

Another solution is to combine the filter with a patcher:

```python
import sys

import loguru
from loguru import logger
from collections import defaultdict

_msg_history = defaultdict(set)


def dedupe(record: "loguru.Record"):
    if "once" in record["extra"]:
        level = record["level"].no
        message = record["message"]
        if message in _msg_history[level]:
            record["extra"]["is_dupe"] = True
        else:
            _msg_history[level].add(message)


def is_not_dupe(record: "loguru.Record") -> bool:
    return not record["extra"].get("is_dupe", False)


logger.remove()
logger.configure(patcher=dedupe)
logger.add(sys.stderr, filter=filter)
logger.add(sys.stderr, filter=filter)

logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message", once=True)
logger.warning("This is a warning message 123", once=True)

logger.info("This is a warning message", once=True)
logger.info("This is a warning message", once=True)
```

However, these approaches have severak limitations. A message containing dynamically formatted placeholder, such as in `logger.info("Value: {}", foo)` will not be deduplicated. In addition, the dictionary can grow indefinitely and appear like a memory leak. 

Another solution is to use an external library such as [`limits`](https://pypi.org/project/limits/) like that:

```python
from loguru import logger
from limits import parse
from limits.storage import MemoryStorage
from limits.strategies import MovingWindowRateLimiter

strategy = MovingWindowRateLimiter(MemoryStorage())
limit = parse("1/second")

while True:
    if strategy.hit(limit):
        logger.info("This message is logged at most once per second.")
```
